### PR TITLE
Use memory map to speed up snapshot untar

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

https://github.com/solana-labs/solana/issues/24798

Use memory map to speed up snapshot untar.

#### Summary of Changes

1. use mmap to read snapshot archive files.
2. retune the parallel factor and chunksize

with mmap
```
reading entire decompressed file took: 44799890 us, bytes: 76654958592, read_us: 23072928, waiting_for_buffer_us: 21715846, largest fetch: 100000000, error: Ok(0)
```

master
```
reading entire decompressed file took: 44960804 us, bytes: 76654958592, read_us: 29821976, waiting_for_buffer_us: 15128175, largest fetch: 100000000, error: Ok(0)
```

The comparison does show that read_time is reduced from 29s to 23 seconds. But the the total untar time don't differ much. The reason is because, although we are reading the data faster, we spend more time on waiting for available buffer, 21s vs 15s. 

Next, we are going to tune the chunk ratio. 

```
// tunable parameters:
// # bytes allocated and populated by reading ahead
const TOTAL_BUFFER_BUDGET_DEFAULT: usize = 2_000_000_000;
// data is read-ahead and saved in chunks of this many bytes
const CHUNK_SIZE_DEFAULT: usize = 100_000_000;
```

```
parallel_factor=8 buf_size=2G chunk=50M
reading entire decompressed file took: 32857423 us, bytes: 76654958592, read_us: 21869626, waiting_for_buffer_us: 10964636, largest fetch: 50000000, error: Ok(0)

parallel_factor=8 buf_size=2G chunk=25M
reading entire decompressed file took: 33978930 us, bytes: 76654958592, read_us: 26325740, waiting_for_buffer_us: 7609328, largest fetch: 25000000, error: Ok(0)
```

So far, best config is *parallel_factor=8 buf_size=2G chunk=50M*. 
untar time reduced from 44s to 32s. 

Fixes #

### CODE REVIEW

File: runtime/src/snapshot_utils.rs [hunk 1]
```diff
@@ -1472,33 +1475,53 @@ fn untar_snapshot_in<P: AsRef<Path>>(
     archive_format: ArchiveFormat,
     parallel_divisions: usize,
 ) -> Result<UnpackedAppendVecMap> {
-    let open_file = || File::open(&snapshot_tar).unwrap();
+    let file = File::open(&snapshot_tar).unwrap();
+
+    // Map the file into immutable memory for better I/O performance
+    let mmap = unsafe { Mmap::map(&file) };
```
Comment (mvines) [inline runtime/src/snapshot_utils.rs:1481]: if this PR makes it out of draft, consider using an existing crate that wraps the unsafe.  https://crates.io/crates/mmarinus seems like one such option

Comment (HaoranYi) [reply runtime/src/snapshot_utils.rs:1481]: It looks like this crate depends on libc. That makes it linux-only. Not for windows.

```
> cargo check
    Checking mmarinus v0.4.0
error[E0433]: failed to resolve: could not find `unix` in `os`
  --> C:\Users\hyi\.cargo\registry\src\github.com-1ecc6299db9ec823\mmarinus-0.4.0\src\builder.rs:11:14
   |
11 | use std::os::unix::io::{AsRawFd, RawFd};
   |              ^^^^ could not find `unix` in `os`
```

Comment (mvines) [reply runtime/src/snapshot_utils.rs:1481]: Perhaps there's another crate, or we could consider upstreaming support for Windows.

Generally it's very undesirable to have raw `unsafe`s in the code like this

Comment (mvines) [reply runtime/src/snapshot_utils.rs:1481]: Alternatively fall back to the original code for Windows?

Comment (HaoranYi) [reply runtime/src/snapshot_utils.rs]: Yeah. Pushed a commit to update this. 
Unfortunately, there is a still unsafe code to convert mmap to slice. This is because the background thread in SharedBufferReader requires the reader to be static. It will requires a bit of rework on the SharedBufferReader, which we can addressed in the future?

File: ledger-tool/src/main.rs [hunk 2]
```diff
@@ -963,6 +963,9 @@ fn main() {
         .validator(is_slot)
         .takes_value(true)
         .help("Halt processing at the given slot");
+    let no_os_memory_stats_reporting_arg = Arg::with_name("no_os_memory_stats_reporting")
```
Comment (HaoranYi) [inline ledger-tool/src/main.rs:966]: This is a different change for adding no_os_memory_stats_reporting cli args to ledger-tool. It is not relevant to the memory map.

File: runtime/src/snapshot_utils.rs [hunk 3]
```diff
@@ -779,7 +779,7 @@ pub struct BankFromArchiveTimings {
 }
 
 // From testing, 4 seems to be a sweet spot for ranges of 60M-360M accounts and 16-64 cores. This may need to be tuned later.
```
Comment (jeffwashington) [inline runtime/src/snapshot_utils.rs:781]: update comment

**Comment by jeffwashington** [comment]
what machine spec did you use to get these metrics?

**Comment by HaoranYi** [comment]
> what machine spec did you use to get these metrics?

gce box: 48core 250G RAM

**Comment by HaoranYi** [comment]
> How big does the slice become? Is it as big as the tar file? Will we run into physical memory limits with 1B/10B account snapshots?
In this test run, it is about 76G. 
Yeah. As long as we have enough *virtual memory* then it will be fine. Otherwise, it will throw an error  and exit.
```
    let mmap = Map::load(&snapshot_tar, Private, perms::Read).unwrap_or_else(|e| {
        error!(
            "Failed to map the snapshot file: {} {}.\n
                        Please increase the virtual memory on the system.",
            snapshot_tar.as_ref().display(),
            e,
        );
        std::process::exit(1);
```
Do you have an estimate of how big the snapshot will be for 1B/10B accounts?

**Comment by HaoranYi** [comment]
Or Perhaps, we can check the file size and switch to old BufReader if the size is too large?

**Comment by jeffwashington** [comment]
> Do you have an estimate of how big the snapshot will be for 1B/10B accounts?

it is compressed, but I imagine it to be basically linear. so if 20G: 170M accounts, then 200G: 1.7B accounts, 2T: 17B accounts. Other things will begin breaking.
I sometimes test on 64G machines. Perhaps that ship has sailed.

**Comment by jeffwashington** [comment]
I'm happy to have this change made. I would love the perf improvements. I load from snapshots all day long.
I am ignorant of how the slice code and the memory mapped file stuff will work in a mem env with very little virtual memory and slices mapped into memory. Does it page the file since it is file backed already and that doesn't count as vm in the same sense? So, will this just work?
Can you write an app to hold all but say 15G of memory on a machine and leave it running. then try ledger-tool verify and see what hapens when we try to untar a 20G snapshot?

**Comment by HaoranYi** [comment]
Yes, will try.

**Comment by sakridge** [comment]
> I'm happy to have this change made. I would love the perf improvements. I load from snapshots all day long.
> I am ignorant of how the slice code and the memory mapped file stuff will work in a mem env with very little virtual memory and slices mapped into memory. Does it page the file since it is file backed already and that doesn't count as vm in the same sense? So, will this just work?
> Can you write an app to hold all but say 15G of memory on a machine and leave it running. then try ledger-tool verify and see what hapens when we try to untar a 20G snapshot?

You mean physical not virtual right? The virtual limit is usually very high like many TB in Linux and you can adjust higher if needed.

**Comment by jeffwashington** [comment]
Well this is in my ignorance. We configure our validators with very little vm (2G)? If we mmap the tar file, then map it into a slice, what happens when we have less physical memory than we need to both represent the mmapped slice AND whatever other dynamic memory the validator needs, like the accounts index. I would assume the os would just page in the mmapped areas that are accessed from the file that backs it. I do not know if this counts in the vm limit or not. And, is the slice 20G elements long? Hopefully this 'just works'.

Basically, can I load a 80G snapshot on a 64G machine with 2G of virtual memory configured?
I suggested using up ram in a separate app so that we can answer the question:
Can I load a 20G snapshot on a machine with 2G of virtual memory and all but ~15G of memory used by another application.

If you know this will work, then I have no concerns when a 2B account snapshot is loaded on a 128G machine.

**Comment by t-nelson** [comment]
it should work, for some definition of "work".  mmap basically "swaps" out to the backing file if there's not enough RAM available.  Whether this is going to have a negative effect on memory pressure for other components is the question.  That is, I'm not sure if the in-memory region is treated as RES or more like VFS cache WRT eviction

**Comment by jeffwashington** [comment]
> Or Perhaps, we can switch to old BufReader if the memory map fails?

I could imagine the mmap succeeds, but it uses almost all of the available mem. Then, we'd oom generating the index or something. I would like generating the index and untarring the append vecs to be happening concurrently - something @HaoranYi and I have been talking about. I'm just trying to think ahead and not lay a trap for ourselves we hit later.

**Comment by jeffwashington** [comment]
> it should work, for some definition of "work". mmap basically "swaps" out to the backing file if there's not enough RAM available. Whether this is going to have a negative effect on memory pressure for other components is the question. That is, I'm not sure if the in-memory region is treated as RES or more like VFS cache WRT eviction

This summarizes my ignorance and concerns ;-) I also don't know. I think it also might 'work'. But I don't know enough low level details of linux memory management and mmapped files to 'know' with confidence. So, I proposed an experiment that seemed pretty easy. Feel free to propose a different experiment!

**Comment by HaoranYi** [comment]
virtual memory can be specified. Usually it is set larger than the physical ram. For example, with 256G RAM, you can set it to vm to 512G. Then you application can map a file that is less than 512G, it should be fine. When you access the data, if it is not in physical ram, os will help to page in data into memory. But it may have bad performance because of thrashing of the available physical ram.

**Comment by jeffwashington** [comment]
> virtual memory can be specified. Usually it is set larger than the physical ram. For example, with 256G RAM, you can set it to vm to 512G. Then you application can map a file that is less than 512G, it should be fine. When you access the data, if it is not in physical ram, os will help to page in data into memory. But it may have bad performance because of thrashing of the available physical ram.

I do not know the history of validators. But, it appears that we set virtual memory at 2G on our machines. I do not know the history of this or the impact if we were to change it.

**Comment by jeffwashington** [comment]
and @HaoranYi if you are testing on a gce machine, I have no idea what VM will be set at by default. I think you're testing on bare metal from the spreadsheet.

**Comment by sakridge** [comment]
The 2G is the swap file that is somewhat different from virtual memory and not really used when mmap'ing a file. Linux will page out to the mapped file in this case and keep recently used pages in physical memory in the page cache. The page cache itself could be maybe be paged out to swap, but swap is not necessary to map a huge vm space and the size of the swap has no bearing on how much you can map. I believe Linux keeps the page cache within the physical memory of the machine.

**Comment by HaoranYi** [comment]
yeah. The vm is set in the following file. In my case, it is 700G. 

```
haoran_yi_solana_com@hyi-dev2:~/git/hy_solana$ cat /etc/sysctl.conf | grep vm.max_map_count
vm.max_map_count=700000
```

**Comment by t-nelson** [comment]
> yeah. The vm is set in the following file on the system. In my case, it is 700G.
> 
> ```
> haoran_yi_solana_com@hyi-dev2:~/git/hy_solana$ cat /etc/sysctl.conf | grep vm.max_map_count
> vm.max_map_count=700000
> ```

That's map count, not byte size.  It's the number of concurrent mmap's allowed

**Comment by HaoranYi** [comment]
> That's map count, not byte size. It's the number of concurrent mmap's allowed

yeah. My current config is 700K. I think the default is only 64K.

**Comment by HaoranYi** [comment]
@jeffwashington 

I did a stress test of using memory when the available physical memory is only
20G. On the machine 378G memory, I have one process consume 360G.
The validator is able to memory map a 84G snapshot file. The actual *physical
memory* increase of during the memory map is about 15G.

**Comment by jeffwashington** [comment]
Ok, so the mmap shouldn't fail due to low memory. Seems good to have confirmed that. It doesn't seem necessary then to fall back on the non-mem map case. Right? When you ran this experiment, did we oom while generating the index?

**Comment by HaoranYi** [comment]
No, it doesn't. I put a std::exit at the end of the load function. It hit
that and terminate.
 Yes, if I let it continues, it will oom when generating the index.

**Review by jeffwashington — APPROVED** [review]
lgtm
